Added view for showing detailed trace statistics

fylip97 opened the pull request:
Signed-off-by: Philip Dengler <philip.dengler@novatec-gmbh.de>

## Which problem is this PR solving?
- This view should help to quickly analyze and identify root causes and anti-patterns. For example, this view can be useful to detect n+1 anti-pattern as @AlexanderWert already mentioned in #1779

![Trace-Overview](https://user-images.githubusercontent.com/44269226/71962084-fdec6580-31f8-11ea-9358-fac5518ae2a7.png)


## Short description of the changes
- A new view called Trace Overview was created. The folder for the view, all dependencies and methods who are needed are located in a subfolder called “TracePageOverview” of “TracePage”

everett980 commented:
Hi Philip, this view looks pretty good but I have a few questions:
1. What are the Service Name and Operation Name drop downs for?
2. What is the color button for?
3. It seems as though ST in Duration is `span max self time / span max time`, I'm wondering if `span max self time / trace duration` might be more useful for time-based RCA? Or in the case of identifying n+1 anti-patterns, `non-overlapping total self time / trace duration` might be what we want? Though that would require calculating the `non-overlapping total self time` for a collection of spans.
4. Also the new folder is called `TraceTagOverview`, is there supposed to be a form of tag analysis? Such as determining which tags have the highest (self) time?

fylip97 commented:
Hi Everett,

thanks for your feedback and questions!

1. The first drop down box specifies the attribute (Service Name, Operation Name or a tag) by which the spans are grouped/splitted. With the second drop down box, it is possible to extend the grouping and split the resulting groups again. For example, if "Service Name" and "Operation Name" are selected, the table splits the spans by the "Service Name" and after that by "Operation Name".

2. The color button colors the table according to the value of the "ST in Duration" with different shades of red. This is done according to the "ST Button" in the "Trace Graph" view.

3. Until now "ST in Duration" is calculated total self time / total duration, like it is done in the "Trace Graph" view.

4. Yes, as mentioned in 1, you can group/split the spans not only by Service and Operation Name, but also use all existing tags here. See also the following example screenshot where I've splitted the spans according to their "database" tag and the "service name". This results in a table which database (if the tag specifies the used database) is used by which service. The "Others" element on the bottom represents the spans which do not contain the splitting attribute, in this case the "database" tag.

![Trace-Overview2](https://user-images.githubusercontent.com/44269226/72059206-dae1b480-32d1-11ea-8b3b-8db0afca3b02.png)

yurishkuro commented:
Add a caption "Group by:" in front of the drop-downs.

Is the first screenshot an example of Colored table?

tiffon commented on packages/jaeger-ui/src/components/TracePage/TraceTagOverview/PopupSQL.css:
```diff
@@ -0,0 +1,67 @@
+/*
```
For consistency (e.g. with `UiFindInput.*`), the file should be named `PopupSql.css`

tiffon commented on packages/jaeger-ui/src/components/TracePage/TraceTagOverview/PopupSQL.tsx:
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2018 The Jaeger Authors.
```
For consistency (e.g. with `UiFindInput.*`), the file should be named `PopupSql.css`

tiffon commented on packages/jaeger-ui/src/components/TracePage/TraceTagOverview/index.css:
```diff
@@ -0,0 +1,59 @@
+limitations under the License.
+*/
+
+#title {
```
Using IDs in CSS selectors is an anti-pattern. What's the advantage of using an ID here instead of a CSS class?

tiffon commented:
Also, can you add a few screenshots showing the various states of the UI?

Also, I meant to add, in your second screenshot ([img](https://user-images.githubusercontent.com/44269226/72059206-dae1b480-32d1-11ea-8b3b-8db0afca3b02.png)), the indentation in the first column looks a like it is not working right.

Lastly, what do you think of also indenting the values in the second column, the count column? I'm wondering if that will make the roll-up / grouping more obvious.

fylip97 commented:
The first screenshot shows the view grouped by only one value.

![1](https://user-images.githubusercontent.com/44269226/72269562-135ff600-3624-11ea-9f00-900a30b33691.png)


The next screenshot shows the possibilities to group them further. Tags are marked with "Tag:".

![2](https://user-images.githubusercontent.com/44269226/72269663-430efe00-3624-11ea-91c3-9022c3565cfe.png)


This should show the interaction with the search. The search is for "api-gateway". The found hits are marked yellow.

![3](https://user-images.githubusercontent.com/44269226/72269712-54f0a100-3624-11ea-8e3c-ee528342f3dc.png)


By clicking on a column, the detailed view below is hidden. In this screen the column with “api-gateway” was clicked and the "Operation Names" for this “Service Name” were hidden.

![4](https://user-images.githubusercontent.com/44269226/72269831-86696c80-3624-11ea-94f0-4b4037ac0d9e.png)


This shows the coloring of the detail view according to the value of the ST in Duration.

![5](https://user-images.githubusercontent.com/44269226/72269921-a731c200-3624-11ea-8505-64a419135c76.png)


At sql tags it is possible to display the complete sql in a popup by clicking on the column.

![7](https://user-images.githubusercontent.com/44269226/72270054-dcd6ab00-3624-11ea-9dbd-2c688aab040a.png)

everett980 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/AltViewOptions.tsx:
```diff
@@ -58,8 +65,12 @@ export default function AltViewOptions(props: Props) {
+      <Button
+        className="ub-mr2"
+        htmlType="button"
+        onClick={() => onTraceGraphViewClicked(selectedTraceView + 1)}
```
> I suggest having this button should show the label of the current view and do nothing when the button itself is clicked.
+1
We had discussed doing this even before this PR. So rather than complicating the button I agree it should only be a dropdown and not a button.

everett980 commented:
I finally got a chance to play around with this branch and I have to say it's very clear how this will be useful!

Now that I've tried it, some follow up questions/feedback:

1. As Yuri mentioned:
> Add a caption "Group by:" in front of the drop-downs.

A caption would be helpful. If you change the `Dropdown`s to [`NameSelector`s](https://github.com/jaegertracing/jaeger-ui/blob/master/packages/jaeger-ui/src/components/DeepDependencies/Header/NameSelector.tsx) you can give the first selector the `label` prop `"Group By"` and still initialize it to "Service Name". The second selector can be given the following props: `placeholder: "Optional Sub-Group..."` and `label: "Sub-Group"`, as well as the `required` prop `false` to make this field clearable.
If you choose to use this component it would be great to move it to `src/components/common`.

2. I would change the "color" button to a [`Checkbox`](https://ant.design/components/checkbox/) next to a `NameSelector` with the `label` "Color by" and the options would be the different numerical columns. This selector should also be required (otherwise a checked `Checkbox` doesn't make much sense) and can default to its current column.

3. Regarding:
> Until now "ST in Duration" is calculated total self time / total duration, like it is done in the "Trace Graph" view.

I'm still not sure how useful % ST in span duration is. A 10s span with 40% ST is a better opportunity for optimization than a 0.2s span with 100% ST but the latter will be much redder.
The "Trace Graph" view is considered experimental and can be tweaked / diverged from as we see fit.
If this is looking to help fix n+1 anti-patterns, then a metric that works out to be "non-concurrent-within-same-parent ST" as well as % of _that_ in total Trace time seems like the silver bullet IMO.
That said, my changes proposed in `2.` would at least address my net ST use case.

4. I'd like to +1 Joe's suggestion:
> Lastly, what do you think of also indenting the values in the second column, the count column? I'm wondering if that will make the roll-up / grouping more obvious.

Indenting sub-group numerical metrics would be nice, especially on large traces where a single group may have *many* sub-groups.

5. I'd prefer if the label "Others" were replaced by another term that conveyed the sub-group dimension was missing. As of yet tags are the only dimension that could be missing, so a label like 'Without tag: "SQL"' should work. This is mostly for the case where there are many sub-groups so "Others" could be misinterpreted as "less prevalent values grouped together", as one would see on a pie chart.

6. This should be considered as a future feature and need not be included yet, but what do you think of the ability to add more than one level of sub-grouping? E.g.: Group by service name & operation name & tag SQL?

7. I don't think the name "Trace Overview" really communicates what this is used for, something like "Trace Statistics" would be more descriptive.

8. Lastly, because I think this feature is great and I would want to promote it if it does not gain traction organically, can you add a GA metric to [TracePageHeader.track.tsx](https://github.com/jaegertracing/jaeger-ui/blob/master/packages/jaeger-ui/src/components/TracePage/TracePageHeader/TracePageHeader.track.tsx) to `trackStatisticsView`?

everett980 commented:
I thought about the following a bit more:
> 3. Regarding:
> 
> > Until now "ST in Duration" is calculated total self time / total duration, like it is done in the "Trace Graph" view.
> 
> I'm still not sure how useful % ST in span duration is. A 10s span with 40% ST is a better opportunity for optimization than a 0.2s span with 100% ST but the latter will be much redder.
> The "Trace Graph" view is considered experimental and can be tweaked / diverged from as we see fit.
> If this is looking to help fix n+1 anti-patterns, then a metric that works out to be "non-concurrent-within-same-parent ST" as well as % of _that_ in total Trace time seems like the silver bullet IMO.
> That said, my changes proposed in `2.` would at least address my net ST use case.

My proposed change in item `2` of my previous comment will also be able to help with n+1 anti-patterns without needing to change or create any metrics. Simply coloring by "Total ST" should do it.

fylip97 commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jaegertracing/jaeger-ui/506)
<!-- Reviewable:end -->

fylip97 commented:
@everett980 do you have any idea why the test "fetchTrace: encountered a declaration exception" is failing?
Locally on my computer all tests run through without problems.

everett980 commented:
> @everett980 do you have any idea why the test "fetchTrace: encountered a declaration exception" is failing?
> Locally on my computer all tests run through without problems.

Looking at the [failing test](https://travis-ci.org/jaegertracing/jaeger-ui/builds/646804840#L896), it seems to use random numbers. It seems to have a 1e-5 chance of failing. Maybe you got unlucky (and maybe I'm misreading the probability, but definitely see `chance`), so I'd recommend running travis again.

yurishkuro commented:
could we not mock random for the tests?

everett980 commented:
> could we not mock random for the tests?

I looks like the code deliberately tries to avoid this issue but has a 1/100000 chance of not avoiding it enough. It should be easy to simply add a small number to one of the two random numbers and it wouldn't happen.

everett980 commented:
@fylip97 I know I wrote a [bit of a saga](https://github.com/jaegertracing/jaeger-ui/pull/506#issuecomment-574772414), but can you let me know your thoughts on these questions?

Also, if the UI has changed in your commits since 1/13, can you please add updated screenshots?

Lastly, many of the comments left by tiffon are now labeled as `Outdated` by Github. This indicates the line was changed, but Github itself doesn't know if the requested changes were implemented. Can you mark completed comments as resolved?

fylip97 commented:
Hi @everett980,
I'd like to respond briefly to your little saga:

1.  I have change the Dropdowns to NameSelectors. I have also added lables before them. The components were moved to src/components/common.

![30](https://user-images.githubusercontent.com/44269226/75354647-9fb13a00-58ad-11ea-9af8-309b3020be2f.png)

![31](https://user-images.githubusercontent.com/44269226/75354798-db4c0400-58ad-11ea-8fd7-e3eac7a68a7b.png)


2. I removed the "color" button and replaced it with a Checkbox. A selector was added next to the Checkbox.

![32](https://user-images.githubusercontent.com/44269226/75354894-fdde1d00-58ad-11ea-984b-6e942984bf82.png)

![33](https://user-images.githubusercontent.com/44269226/75354951-12221a00-58ae-11ea-9bd1-66f0fd63af46.png)


3. I change the calculation. The ST in Duration is now the percentage of ST of the total trace duration.

4. All values from the Sub Group are now indented.

5. I have change the Other label to Without Tag.

6. The view is now called Trace Statistics.

7. GA metrics are added.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/AltViewOptions.tsx:
```diff
@@ -16,29 +16,62 @@ import * as React from 'react';
+  viewType: ETraceViewType;
 };
 
+const menuItems = [
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/AltViewOptions.tsx:
```diff
@@ -61,10 +94,14 @@ export default function AltViewOptions(props: Props) {
     <Dropdown overlay={menu}>
-      <Button className="ub-mr2" htmlType="button" onClick={handleToggleView}>
-        Alternate Views <Icon type="down" />
+      <Button className="ub-mr2" htmlType="button">
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/DetailTableData.css:
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright (c) 2018 The Jaeger Authors.
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/DetailTableData.test.js:
```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2018 The Jaeger Authors.
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/HeaderTable.tsx:
```diff
@@ -0,0 +1,50 @@
+};
+
+/**
+ * Used to render the table header.
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/DetailTableData.tsx:
```diff
@@ -0,0 +1,108 @@
+  type: string;
+  name: string;
+  searchColor: string;
+  values: any[];
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/DetailTableData.tsx:
```diff
@@ -0,0 +1,108 @@
+  name: string;
+  searchColor: string;
+  values: any[];
+  columnsArray: any[];
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/HeaderTable.css:
```diff
@@ -0,0 +1,32 @@
+  background: rgb(236, 236, 236);
+  border: 1px solid rgb(204, 204, 204);
+  font-size: 100%;
+  font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;
```
Done. 
Lines were removed.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/index.tsx:
```diff
@@ -274,12 +275,14 @@ export class TracePageImpl extends React.PureComponent<TProps, TState> {
 
-  toggleTraceGraphView = () => {
-    const { traceGraphView } = this.state;
+  toogleTraceView = (actualViewType: ETraceViewType) => {
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/index.tsx:
```diff
@@ -274,12 +275,14 @@ export class TracePageImpl extends React.PureComponent<TProps, TState> {
+  toogleTraceView = (actualViewType: ETraceViewType) => {
+    let { viewType } = this.state;
+
     if (this.props.trace && this.props.trace.data) {
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/index.tsx:
```diff
@@ -334,10 +337,10 @@ export class TracePageImpl extends React.PureComponent<TProps, TState> {
+    let spanFindMatches: Set<string> | null | undefined;
     if (uiFind) {
-      if (traceGraphView) {
+      if (viewType !== ETraceViewType.TraceTimelineViewer) {
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/AltViewOptions.tsx:
```diff
@@ -16,29 +16,62 @@ import * as React from 'react';
-    onTraceGraphViewClicked();
+  const { onTraceViewChange, viewType, traceID } = props;
+
+  const handleToggleView = (item: ETraceViewType) => {
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/AltViewOptions.tsx:
```diff
@@ -61,10 +94,14 @@ export default function AltViewOptions(props: Props) {
-      <Button className="ub-mr2" htmlType="button" onClick={handleToggleView}>
-        Alternate Views <Icon type="down" />
+      <Button className="ub-mr2" htmlType="button">
+        {menuItems.find(test => test.viewType === viewType) !== undefined
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/AltViewOptions.tsx:
```diff
@@ -16,29 +16,62 @@ import * as React from 'react';
-          {traceGraphView ? 'Trace Timeline' : 'Trace Graph'}
-        </a>
-      </Menu.Item>
+      {menuItems.map(item =>
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/TraceStatisticsHeader.tsx:
```diff
@@ -0,0 +1,247 @@
+  ['Avg', 'avg'],
+  ['Min', 'min'],
+  ['Max', 'max'],
+  ['Total ST', 'self'],
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/TraceStatisticsHeader.tsx:
```diff
@@ -0,0 +1,247 @@
+  ['ST Avg', 'selfAvg'],
+  ['ST Min', 'selfMin'],
+  ['ST Max', 'selfMax'],
+  ['ST Duration', 'percent'],
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/DetailTableData.css:
```diff
@@ -0,0 +1,54 @@
+  padding-left: 0.6em;
+}
+
+.DetailTableData--td {
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/HeaderTable.tsx:
```diff
@@ -0,0 +1,50 @@
+ * @param props
+ */
+export default function HeaderTable(props: Props) {
+  const thStyle = { width: Math.round(window.innerWidth * 0.2) };
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/HeaderTable.css:
```diff
@@ -0,0 +1,32 @@
+}
+
+.HeaderTable--buttonPosition {
+  float: right;
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TraceStatistics/TraceStatisticsHeader.tsx:
```diff
@@ -0,0 +1,247 @@
+   */
+  clearValue() {
+    this.setState({
+      valueNameSelector2: 'No Item selected',
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/index.tsx:
```diff
@@ -274,12 +275,14 @@ export class TracePageImpl extends React.PureComponent<TProps, TState> {
 
-  toggleTraceGraphView = () => {
-    const { traceGraphView } = this.state;
+  toogleTraceView = (actualViewType: ETraceViewType) => {
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/index.tsx:
```diff
@@ -384,28 +389,41 @@ export class TracePageImpl extends React.PureComponent<TProps, TState> {
+              default:
+                return (
+                  <section style={{ paddingTop: headerHeight }}>
+                    <TraceStatistics trace={data} uiFindVertexKeys={graphFindMatches} uiFind={uiFind} />
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/index.test.js:
```diff
@@ -628,16 +629,16 @@ describe('<TracePage>', () => {
     });
 
     it('propagates traceGraphView changes', () => {
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/AltViewOptions.tsx:
```diff
@@ -16,29 +16,62 @@ import * as React from 'react';
 type Props = {
-  onTraceGraphViewClicked: () => void;
-  traceGraphView: boolean;
+  onTraceViewChange: (actualViewType: ETraceViewType) => void;
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/TracePageHeader.tsx:
```diff
@@ -49,7 +49,7 @@ type TracePageHeaderEmbedProps = {
   onArchiveClicked: () => void;
   onSlimViewClicked: () => void;
-  onTraceGraphViewClicked: () => void;
+  onTraceViewChange: (actualViewType: ETraceViewType) => void;
```
Done.

fylip97 commented on packages/jaeger-ui/src/components/TracePage/index.test.js:
```diff
@@ -628,16 +629,16 @@ describe('<TracePage>', () => {
       wrapper.update();
       refreshWrappers();
-      expect(header.prop('traceGraphView')).toBe(true);
+      expect(header.prop('viewType')).toBe(undefined);
```
Done.

everett980 commented:
Hi @fylip97, just wanted to check in on this PR. About a month ago you replied to nearly all of my feedback saying they've been addressed. If you'd like help with the last two items or don't have time to complete that yourself, you can push the changes you have done and I can review the changes and tackle the last two items.

fylip97 commented:
Hi @everett980 
I just pushed my changes.
At the moment I have very little time and would like to accept your offer.

idohalevi commented:
@fylip97 @everett980 Thank you for this great feature. Is there any way I can assist in pushing this PR?

yurishkuro commented:
🎉 🎉 🎉

AlexanderWert commented:
Great! Happy to see this feature being approved!
